packages/toolpad-app/src/toolpad/Home.tsx

```diff
@@ -77,7 +93,12 @@ function CreateAppDialog({ onClose, ...props }: CreateAppDialogProps) {
         <DialogForm
           onSubmit={(event) => {
             event.preventDefault();
-            createAppMutation.mutate([name, { dom: dom.trim() ? JSON.parse(dom) : null }]);
+
+            const selectedAppTemplateDom =
+              APP_TEMPLATE_OPTIONS.find(({ value }) => value === appTemplate)?.dom || null;
+            const appDom = dom.trim() ? JSON.parse(dom) : selectedAppTemplateDom;
+
+            createAppMutation.mutate([name, { dom: appDom }]);
```

Comment:
I feel like these templates should live on the backend. So that we don't pull all of them in the frontend bundles. This will definitely become problematic as we add more of them.

I've been wondering whether we should update the `createApp` method and give it a `from` option, which would contain a discriminated union specifying what we want to start from. Either a provided dom, or a template:

```tsx
export interface CreateAppOptions {
  from?:
    | {
        kind: 'dom';
        dom: appDom.AppDom;
      }
    | {
        kind: 'template';
        id: string;
      }
    | null;
}

export async function createApp(name: string, opts: CreateAppOptions = {}): Promise<prisma.App>;
```

@bharatkashyap This could also incorporate the app duplication if we add
```tsx
    | {
        kind: 'app';
        id: string;
      }
```
to the discriminated union.

Reply:
> @bharatkashyap This could also incorporate the app duplication if we add
> 
> ```tsx
>     | {
>         kind: 'app';
>         id: string;
>       }
> ```
> 
> to the discriminated union.

I feel there's a difference between a `kind: 'dom'` and every other type of `kind`, since `template` and `app` are fetching  a `dom` and then passing them into `createApp`.

In that sense, the `kind`s do not seem to be on the same level of abstraction. For me, it is okay for `createApp` with `dom` to continue being the lowest-level abstraction. 

A `from` option might suit a `fetchDom` function which should be callable inside wrapper functions we create to duplicate apps, create them from templates and so on.

So, 
```ts
export async function duplicateApp(id: string, name: string): Promise<AppMeta> {
  const dom = await fetchDom({ from: { kind: 'app', id: id}})
  const newApp = await createApp(`${name} (copy)`, { dom });
  return newApp;
}
```

and,
```ts
export async function createTemplateApp(id: string, name: string): Promise<AppMeta> {
  const dom = await fetchDom({ from: { kind: 'template', id: id}})
  const newApp = await createApp(name, { dom });
  return newApp;
}
```

Reply:
> In that sense, the kinds do not seem to be on the same level of abstraction.

They're all creating an new application based on an initializer. the initializer can be another "app", a predefined "template", or an inline "dom". That's the proposed abstraction.

As for `fetchDom`, I think we should figure out the right abstractions as the need arises. I think we're starting to see that DOM on its own doesn't have a lot of meaning. And it will have less so in the context of a template when we make them parametrizable.

Reply:
> > In that sense, the kinds do not seem to be on the same level of abstraction.
> 
> They're all creating a new application based on an initializer. the initializer can be another "app", a predefined "template", or an inline "dom". That's the proposed abstraction.
> 
Would you not say that the predefined "template" and another "app" are essentially an abstraction over fetching the dom attached to the respective template/app? That's the only issue in my head.

Reply:
An "abstraction" hides implementation details. Would you not say that the method by which a dom is obtained is an implementation detail? 
Suppose some day we want to add a way to create an app from JSON hosted under a remote url. Wouldn't that fit under the same abstraction?

Reply:
> Suppose some day we want to add a way to create an app from JSON hosted under a remote url. Wouldn't that fit under the same abstraction?

Yep, I agree it would. To me it would fit under another kind of "fetching the DOM" (via URL) and not "creating an app" (which is always through a dom). Even if none is passed, we generate a "default" dom. 

In essence, I'm trying to say that the "kinds" in the example fit better as implementation details of fetching the DOM and not creating an app. 

In the future, if there starts to exist another way of creating an app (say integrating with an existing codebase, which doesn't require generating a dom) that would feel more fitting as another "kind" of `createApp`.

Reply:
I don't think any of the 2 prosposals are bad, but a single `createApp` sounds a bit easier to extend/mantain in the long run in my opinion, in case we end up having many different ways of creating apps. i'll start with that and move the templates to the backend.
Shouldn't be too hard to adjust it in the future if we see the need to.